From c7daeb889bbadbaeca613886b09558a643470541 Mon Sep 17 00:00:00 2001
From: fraz0815 <l.schimmel83@gmail.com>
Date: Wed, 30 Jul 2025 17:36:53 +0200
Subject: [PATCH 1/2] fix for 6.16

Rename functions:
`rdmsrl_safe_on_cpu` -> `rdmsrq_safe_on_cpu`

`rdmsrl_safe` -> `rdmsrq_safe`
---
 zenergy.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/zenergy.c b/zenergy.c
index 0a66df6..4b15fb0 100644
--- a/zenergy.c
+++ b/zenergy.c
@@ -80,7 +80,7 @@ static void get_energy_units(struct zenergy_data *data)
 {
 	u64 rapl_units;
 
-	rdmsrl_safe(ENERGY_PWR_UNIT_MSR, &rapl_units);
+	rdmsrq_safe(ENERGY_PWR_UNIT_MSR, &rapl_units);
 	data->energy_units = (rapl_units & zenergy_UNIT_MASK) >> 8;
 }
 
@@ -89,7 +89,7 @@ static void __accumulate_delta(struct sensor_accumulator *accum,
 {
 	u64 input;
 
-	rdmsrl_safe_on_cpu(cpu, reg, &input);
+	rdmsrq_safe_on_cpu(cpu, reg, &input);
 	input &= zenergy_MASK;
 
 	if (input >= accum->prev_value)

From 383b0b91160bdd12fe13ad263355b0a7e9c012a5 Mon Sep 17 00:00:00 2001
From: fraz0815 <l.schimmel83@gmail.com>
Date: Thu, 31 Jul 2025 14:08:26 +0200
Subject: [PATCH 2/2] check kernel version

.. before using new functions and not break old stuff
---
 zenergy.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/zenergy.c b/zenergy.c
index 4b15fb0..c3f495e 100644
--- a/zenergy.c
+++ b/zenergy.c
@@ -80,7 +80,12 @@ static void get_energy_units(struct zenergy_data *data)
 {
 	u64 rapl_units;
 
+	#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 16, 0)
+	rdmsrl_safe(ENERGY_PWR_UNIT_MSR, &rapl_units);
+	#else
 	rdmsrq_safe(ENERGY_PWR_UNIT_MSR, &rapl_units);
+	#endif
+	
 	data->energy_units = (rapl_units & zenergy_UNIT_MASK) >> 8;
 }
 
@@ -88,8 +93,11 @@ static void __accumulate_delta(struct sensor_accumulator *accum,
 			       int cpu, u32 reg)
 {
 	u64 input;
-
+	#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 16, 0)
+	rdmsrl_safe_on_cpu(cpu, reg, &input);
+	#else
 	rdmsrq_safe_on_cpu(cpu, reg, &input);
+	#endif
 	input &= zenergy_MASK;
 
 	if (input >= accum->prev_value)
